src/codegen/compile.ml

```diff
@@ -10646,126 +10494,70 @@ and compile_prim_invocation (env : E.t) ae p es at =
       get_b ^^
       BitTagged.if_tagged_scalar env [F64Type]
         (get_b ^^
-         BitTagged.untag_i32 __LINE__ env Type.Int ^^
-         G.i (Convert (Wasm.Values.I64 I64Op.ExtendSI32)) ^^
-         G.i (Convert (Wasm.Values.F64 F64Op.ConvertSI64)))
+         BitTagged.untag __LINE__ env Type.Int ^^
+         G.i (Convert (Wasm_exts.Values.F64 F64Op.ConvertSI64)))
         (get_b ^^
          E.call_import env "rts" "bigint_to_float64")
 
     | Float, Int64 ->
       SR.UnboxedWord64 Int64,
       compile_exp_as env ae SR.UnboxedFloat64 e ^^
-      G.i (Convert (Wasm.Values.I64 I64Op.TruncSF64))
+      G.i (Convert (Wasm_exts.Values.I64 I64Op.TruncSF64))
 
     | Int64, Float ->
       SR.UnboxedFloat64,
       compile_exp_as env ae (SR.UnboxedWord64 Int64) e ^^
-      G.i (Convert (Wasm.Values.F64 F64Op.ConvertSI64))
+      G.i (Convert (Wasm_exts.Values.F64 F64Op.ConvertSI64))
     | Nat8, Nat16 ->
-      SR.UnboxedWord32 Nat16,
-      compile_exp_as env ae (SR.UnboxedWord32 Nat8) e ^^
+      SR.UnboxedWord64 Nat16,
+      compile_exp_as env ae (SR.UnboxedWord64 Nat8) e ^^
       TaggedSmallWord.lsb_adjust Nat8 ^^
       TaggedSmallWord.msb_adjust Nat16
     | Nat16, Nat32 ->
-      SR.UnboxedWord32 Nat32,
-      compile_exp_as env ae (SR.UnboxedWord32 Nat16) e ^^
-      TaggedSmallWord.lsb_adjust Type.Nat16 ^^
-      TaggedSmallWord.msb_adjust Nat32 (* NB: a nop for 32-bit present, but not for 64-bit future *)
+      SR.UnboxedWord64 Nat32,
+      compile_exp_as env ae (SR.UnboxedWord64 Nat16) e ^^
+      TaggedSmallWord.lsb_adjust Nat16 ^^
+      TaggedSmallWord.msb_adjust Nat32
     | Nat32, Nat64 ->
       SR.UnboxedWord64 Nat64,
-      compile_exp_as env ae (SR.UnboxedWord32 Nat32) e ^^
-      G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32))
-    | Nat16, (Nat8 as pty) ->
-      SR.UnboxedWord32 Nat8,
-      let num_bits = (TaggedSmallWord.bits_of_type pty) in
-      let set_val, get_val = new_local env "convertee" in
-      compile_exp_as env ae (SR.UnboxedWord32 Nat16) e ^^
-      set_val ^^
-      get_val ^^
-      compile_shrU_const (Int32.of_int (32 - num_bits)) ^^
-      E.then_trap_with env "losing precision" ^^
-      get_val ^^
-      compile_shl_const (Int32.of_int num_bits)
-    | Nat32, (Nat16 as pty) ->
-      SR.UnboxedWord32 pty, (* ! *)
-      let num_bits = Int32.of_int (TaggedSmallWord.bits_of_type pty) in
+      compile_exp_as env ae (SR.UnboxedWord64 Nat32) e ^^
+      TaggedSmallWord.lsb_adjust Nat32
```

Comment:
I guess these three cases could also be refactor into one using .. as from_type and .. as typ as you do very nicely for the other conversions below.

Reply:
Thanks for the suggestion. I have refactored this.